File: tasks/implementations/php/tasks/binary_search_tree.php
<?php
class BSTNode {
  public $value;
  public $left = null;
  public $right = null;
  public function __construct(int $value) { $this->value = $value; }
}

function bst_insert(?BSTNode $root, int $value): BSTNode {
  if ($root === null) return new BSTNode($value);
  $node = $root;
  while (true) {
    if ($value < $node->value) {
      if ($node->left === null) { $node->left = new BSTNode($value); return $root; }
      $node = $node->left;
    } elseif ($value > $node->value) {
      if ($node->right === null) { $node->right = new BSTNode($value); return $root; }
      $node = $node->right;
    } else {
      return $root;
    }
  }
}

function bst_contains(?BSTNode $root, int $value): bool {
  $node = $root;
  while ($node !== null) {
    if ($value < $node->value) $node = $node->left;
    elseif ($value > $node->value) $node = $node->right;
    else return true;
  }
  return false;
}

function run_binary_search_tree(string $size, string $fixturesRoot): void {
  $tokens = array_map('intval', preg_split('/\s+/', trim(file_get_contents($fixturesRoot . "/generated/bst/{$size}.txt"))));
  $insertCount = $tokens[0];
  $inserts = array_slice($tokens, 1, $insertCount);
  $queryIndex = 1 + $insertCount;
  $queryCount = $tokens[$queryIndex];
  $queries = array_slice($tokens, $queryIndex + 1, $queryCount);
  $root = null;
  foreach ($inserts as $value) $root = bst_insert($root, $value);
  $total = 0;
  foreach ($queries as $value) if (bst_contains($root, $value)) $total += $value;
  echo $total . PHP_EOL;
}

File: tasks/implementations/go/tasks/binary_search_tree.go
package main

import (
    "fmt"
    "os"
    "path/filepath"
    "strconv"
    "strings"
)

type BSTNode struct {
    value int
    left  *BSTNode
    right *BSTNode
}

func bstInsert(root *BSTNode, value int) *BSTNode {
    if root == nil { return &BSTNode{value: value} }
    node := root
    for {
        if value < node.value {
            if node.left == nil { node.left = &BSTNode{value: value}; return root }
            node = node.left
        } else if value > node.value {
            if node.right == nil { node.right = &BSTNode{value: value}; return root }
            node = node.right
        } else {
            return root
        }
    }
}

func bstContains(root *BSTNode, value int) bool {
    node := root
    for node != nil {
        if value < node.value { node = node.left } else if value > node.value { node = node.right } else { return true }
    }
    return false
}

func RunBinarySearchTree(size string, fixturesRoot string) {
    text, _ := os.ReadFile(filepath.Join(fixturesRoot, "generated", "bst", size+".txt"))
    tokens := strings.Fields(string(text))
    insertCount, _ := strconv.Atoi(tokens[0])
    var root *BSTNode
    for i := 0; i < insertCount; i++ {
        value, _ := strconv.Atoi(tokens[1+i])
        root = bstInsert(root, value)
    }
    queryIndex := 1 + insertCount
    queryCount, _ := strconv.Atoi(tokens[queryIndex])
    total := 0
    for i := 0; i < queryCount; i++ {
        value, _ := strconv.Atoi(tokens[queryIndex+1+i])
        if bstContains(root, value) { total += value }
    }
    fmt.Println(total)
}
